src/d3a/models/strategy/external_strategies/pv.py

```diff
@@ -17,21 +17,33 @@
 """
 import json
 import logging
-from collections import deque
-from typing import Dict
+from typing import Dict, Callable, TYPE_CHECKING
 
 from d3a_interface.constants_limits import ConstSettings
 from d3a_interface.data_classes import Offer
-from pendulum import duration, DateTime
+from pendulum import duration
 
 from d3a.d3a_core.util import get_market_maker_rate_from_config
 from d3a.models.strategy.external_strategies import (
     ExternalMixin, IncomingRequest, ExternalStrategyConnectionManager, default_market_info)
+from d3a.models.strategy.external_strategies.forecast_mixin import ForecastExternalMixin
 from d3a.models.strategy.predefined_pv import PVPredefinedStrategy, PVUserProfileStrategy
 from d3a.models.strategy.pv import PVStrategy
 
+if TYPE_CHECKING:
+    from d3a.models.state import PVState
+    from d3a.models.strategy import Offers
+
 
 class PVExternalMixin(ExternalMixin):
+
+    state: "PVState"
+    offers: "Offers"
+    can_offer_be_posted: Callable
+    post_offer: Callable
+    set_produced_energy_forecast_kWh_future_markets: Callable
+    _delete_past_state: Callable
```

Comment:
Just to rephrase what you wrote in the description of the PR:
With these type hints you are making sure that the `PVExternalMixin` is aware of these methods and even if the `ExternalMixin` is not implementing them, no error is raised by pycharm, right?

Reply:
Precisely, but more importantly, the future developer will know that these are the methods that need to be implemented by the mixed-in class. It is probably an overkill for the user of the class since we all more or less know that this class needs to be mixed-in to a subclass of BaseStrategy/BidEnabledStrategy, however for the future maintainer it might facilitate him when refactoring this class.